### PR TITLE
Volodin Evgeniy. Lab №3 Option №1.

### PR DESCRIPTION
**Pull request by evldn**
In this lab, a pass was implemented that replaces the argument-dependent mul and add instructions with one - muladd.
**The source code:**
``` C++
#include <immintrin.h>

__m128d simple_test(__m128d a, __m128d b, __m128d c, __m128d d) {
	__m128d temp = a * b;
	d = temp + c;
	return d;
}
```
**LLVM MIR code before applying the pass:**
``` MIR
body:             |
  bb.0.entry:
    liveins: $xmm0, $xmm1, $xmm2
  
    renamable $xmm0 = nofpexcept MULPDrr killed renamable $xmm0, killed renamable $xmm1, implicit $mxcsr
    renamable $xmm0 = nofpexcept ADDPDrr killed renamable $xmm0, killed renamable $xmm2, implicit $mxcsr
    RET64 $xmm0
```
**LLVM MIR code after applying the pass:**
``` MIR
body:             |
  bb.0.entry:
    liveins: $xmm0, $xmm1, $xmm2
  
    $xmm0 = VFMADD213PDr $xmm0, $xmm1, $xmm2, implicit $mxcsr
    RET64 $xmm0
```

### CODE REVIEW

File: llvm/lib/Target/X86/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,7 @@ tablegen(LLVM X86GenFoldTables.inc -gen-x86-fold-tables -asmwriternum=1)
 add_public_tablegen_target(X86CommonTableGen)
 
 set(sources
+  X86VolodinEMulAddPass.cpp
```
Comment (m-ly4) [inline llvm/lib/Target/X86/CMakeLists.txt]: Please rewrite it as shared lib plugin.

File: llvm/lib/Target/X86/X86VolodinEMulAddPass.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+INITIALIZE_PASS(X86VolodinEMulAddPass, NAME, DESC, false, false)
+
+bool findInstruction(std::vector<MachineInstr *> &MIvector,
+                     MachineInstr *instruction) {
+  auto result{std::find(begin(MIvector), end(MIvector), instruction)};
+  if (result == end(MIvector))
+    return false;
+  else
+    return true;
```
Comment (m-ly4) [inline llvm/lib/Target/X86/X86VolodinEMulAddPass.cpp]:
```suggestion
  return (result != end(MIvector));
```

File: llvm/lib/Target/X86/X86VolodinEMulAddPass.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+bool X86VolodinEMulAddPass::runOnMachineFunction(MachineFunction &MF) {
+  const X86Subtarget &STI = MF.getSubtarget<X86Subtarget>();
+  const X86InstrInfo &TII = *STI.getInstrInfo();
+
+  bool Changed = false;
+  std::vector<MachineInstr *> MIvector;
+
+  for (MachineBasicBlock &MBB : MF) {
+    for (auto iterator = MBB.begin(); iterator != MBB.end(); ++iterator) {
+      if (iterator->getOpcode() == X86::MULPDrr) {
```
Comment (m-ly4) [inline llvm/lib/Target/X86/X86VolodinEMulAddPass.cpp]:
```suggestion
      if (iterator->getOpcode() != X86::MULPDrr) {
        continue;
      }
```

File: llvm/test/lab3/test_muladd.mir [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+# 	d = t + c;
+# 	return d / b;
+# }
+# 
+# __m128d test_without_muladd(__m128d a, __m128d b, __m128d c, __m128d d, __m128d e) {
+# 	d = a * b;
+# 	e = a + c;
+# 	return d - e;
+# }
+
```
Comment (m-ly4) [inline llvm/test/lab3/test_muladd.mir]: Please add also:
```
# __m128d another_test(__m128d a, __m128d b, __m128d c, __m128d d) {
# 	__m128d t = a * b;
# 	__m128d t2 = c * d;
# 	t2 = t + t2;
# 	t = t + c;
# 	return t * t2;
# }
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by evldn** [comment]
1. I rewrote pass as shared lib plugin.
2. I added the test you suggested, which was performed incorrectly on the old version of the pass. I fixed this by adding additional checks to the pass that track the mutability of operands between the mul and add instructions.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**